GpsDataCapturer/sharedlibrary/src/main/java/com/google/sharedlibrary/GpxWriteHandler.java

```diff
@@ -0,0 +1,121 @@
+ */
+public class GpxWriteHandler implements Runnable {
+    private static final String TAG = "GpxWriterHandler";
+    private final String formattedTime;
+    private final Location location;
+    private final File gpxFile;
+    private final Context context;
+    private final boolean append;
+    private boolean isNewFile;
+    private static final int SIZE = 20480;
+
+
+    public GpxWriteHandler(Context context, String formattedTime, File gpxFile,
+            Location location, boolean append, boolean isNewFile) {
+        this.context = context;
+        this.formattedTime = formattedTime;
+        this.gpxFile = gpxFile;
+        this.location = location;
+        this.append = append;
+        this.isNewFile = isNewFile;
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.O)
+    @Override
+    public void run() {
+        Log.i(TAG, "Start writing to file");
+        try (FileWriter fileWriter = new FileWriter(gpxFile, true)) {
+            BufferedWriter bufferedWriter = new BufferedWriter(fileWriter, SIZE);
+
+            //Write the header if isNewFile
+            if (isNewFile) {
+                Log.d(TAG, "Writing new file header.");
+                bufferedWriter.write(fileHeader(Utils.getFormattedCurrentTime(context)));
+            }
+
+            //write the captured gps data to file
+            bufferedWriter.write(getTrackPointXml(location, formattedTime));
+            bufferedWriter.flush();
+            bufferedWriter.close();
+            fileWriter.close();
+            Log.d(TAG, getTrackPointXml(location, formattedTime));
+        } catch (Exception e) {
+            Log.e(TAG, "GpxFileWriter.write", e);
+        }
+    }
+
+    /**
+     * Create the xml header with version, creator and metadata
+     *
+     * @param formattedStartTime time of on location changed in format
+     * @return A header string
+     */
+    public static String fileHeader(String formattedStartTime) {
+        StringBuilder header = new StringBuilder();
+
+        header.append("<?xml version='1.0' encoding='UTF-8' ?>");
+        header.append("<gpx version=\"1.1\" creator=\"GpsDataCapturer " + BuildConfig.VERSION_CODE
+                + "\" ");
+        header.append("xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" ");
+        header.append("xmlns=\"http://www.topografix.com/GPX/1/1\" ");
+        header.append("xsi:schemaLocation=\"http://www.topografix.com/GPX/1/1 ");
+        header.append("http://www.topografix.com/GPX/1/1/gpx.xsd\">\n");
+        header.append("<metadata><time>").append(formattedStartTime).append("</time>");
+        header.append("<device>").append(Build.DEVICE).append("</device>");
+        header.append("<id>").append(Build.ID).append("</id>");
+        header.append("<manufacturer>").append(Build.MANUFACTURER).append("</manufacturer>");
+        header.append("<model>").append(Build.MODEL).append("</model></metadata>\n");
+        header.append("<trk>");
+        header.append("<trkseg>");
```

Comment:
Do we need new lines for these 2?

Reply:
Yes, it's needed corresponding to the footer </trkseg></trk>. Or are you suggesting that we don't need to store the tracking points to a segment nor a track? If so, we don't need them.

Reply:
Done.